lib/phoenix/controller/pipeline.ex

```diff
@@ -125,11 +126,40 @@ defmodule Phoenix.Controller.Pipeline do
         _ = var!(controller)
         _ = var!(action)
 
-        unquote(body)
+        try do
+          unquote(body)
+        catch
+          kind, reason ->
+            Phoenix.Controller.Pipeline.__catch__(
+              var!(conn), kind, reason, System.stacktrace, var!(controller), var!(action)
+            )
+        end
       end
     end
   end
 
+  @doc false
+  def __catch__(conn, :error = kind, :function_clause = reason, stack, controller, action) do
+    case stack do
+      [{^controller, ^action, [%Plug.Conn{}, %{} | _], location} | _] ->
```

Comment:
I would check only the first argument. :) Also, I think you could just raise a Phoenix specific error in this clause, something like:

```elixir
defmodule Phoenix.ActionClauseError do
  defexception [conn: ..., plug_status: 400, controller: nil, action: nil]

  # ...
end
```

And let the error layer take care of it. This way  you don't need to log things like file, line and stacktrace and parameters. The error layer will take care of it for us.